client/core/trade.go

```diff
@@ -327,132 +322,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
 		if oid != t.ID() {
 			return fmt.Errorf("negotiate called for wrong order. %s != %s", oid, t.ID())
 		}
-		// Contract txn asset: buy means quote, sell means base. NOTE:
-		// msgjson.Match could instead have just FeeRateSwap for the recipient,
-		// but the other fee rate this way could be of value for auditing the
-		// counter party's contract txn.
-		feeRateSwap := msgMatch.FeeRateQuote
-		if trade.Sell {
-			feeRateSwap = msgMatch.FeeRateBase
+
+		// First check if this is a match with a cancel order, in which case the
+		// counterparty Address field would be empty. If the user placed a
+		// cancel order, that order will be recorded in t.cancel on cancel order
+		// creation via (*dexConnection).tryCancel or restore from DB via
+		// (*Core).dbTrackers.
+		if t.cancel != nil && msgMatch.Address == "" {
+			log.Infof("Maker notification for cancel order received for order %s. match id = %s",
+				oid, msgMatch.MatchID)
+			cancelMatch = msgMatch
+			newFill += msgMatch.Quantity // should push this trade to 100% filled
+			continue
 		}
-		log.Infof("Starting negotiation for match %v for order %v with feeRateSwap = %v",
-			mid, oid, feeRateSwap)
+
 		match := &matchTracker{
-			id:     mid,
-			prefix: t.Prefix(),
-			trade:  trade,
-			MetaMatch: db.MetaMatch{
-				MetaData: &db.MatchMetaData{
-					Status: order.NewlyMatched,
-					Proof: db.MatchProof{
-						Auth: db.MatchAuth{
-							MatchSig:   msgMatch.Sig,
-							MatchStamp: msgMatch.ServerTime,
-						},
-					},
-					DEX:   t.dc.acct.host,
-					Base:  t.Base(),
-					Quote: t.Quote(),
-				},
-				Match: &order.UserMatch{
-					OrderID:     oid,
-					MatchID:     mid,
-					Quantity:    msgMatch.Quantity,
-					Rate:        msgMatch.Rate,
-					Address:     msgMatch.Address,
-					Status:      order.MatchStatus(msgMatch.Status),
```

Comment:
Note that this previous code was potentially setting different Status values between the MatchMetaData and UserMatch.  Since I `negotiate` is intended to start new matches, I believe, I've updated this with a `SetStatus(order.NewlyMatched)` several lines down, while the new `makeMetaMatch` method ensures both Status values are consistent.